Zephyr Shared ISRs

sambhurst opened the pull request:
This PR implements a feature that allows for the use of shared interrupts in Zephyr. I've only tested it on the arm aarch32. Please see #49379 for more context. Also, I'm not a python programmer, so I'm sure my changes to gen_isr_tables.py can be improved upon.

sambhurst committed:
scripts: Enable gen_isr_tables.py to generate shared IRQs

Adds a feature to gen_isr_tables.py that enables it to
detect and generate shared IRQs when IRQ_CONNECT is used.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

microbuilder commented on include/zephyr/irq.h:
```diff
@@ -111,6 +111,51 @@ irq_connect_dynamic(unsigned int irq, unsigned int priority,
+ * must be subsequently enabled via irq_enable() before the interrupt handler
+ * begins servicing interrupts.
+ *
+ * Up to three ISRs can share one IRQ and they are called in sequence but
```
The Kconfig file states `This option allows up to 4 ISRs to share 1 IRQ line` ... presumably this should be four (`MAX_SHARED_ISRS = 4`).

microbuilder commented on include/zephyr/sw_isr_table.h:
```diff
@@ -64,6 +64,12 @@ struct _isr_list {
 
+/**
+ * This interrupt gets put directly in the vector table and can
+ * be shared with up to three ISRs
```
The Kconfig file states `This option allows up to 4 ISRs to share 1 IRQ line` ... presumably this should be four (`MAX_SHARED_ISRS = 4`).

sambhurst replied:
I've removed the max number of shared ISRs. Any number of ISRs can been shared.

microbuilder commented on scripts/build/gen_isr_tables.py:
```diff
@@ -14,6 +14,9 @@
 ISR_FLAG_DIRECT = (1 << 0)
+ISR_FLAG_SHARE  = (1 << 1)
+
+MAX_SHARED_ISRS = 4
```
Is there a specific reason for this particular number? It might be worth putting a justification in the comment, or describe tradeoffs of making it larger or smaller.

andyross replied:
More nitpicky still: why put a limit in?  Both a singly-linked list or a null-terminated array can be generated at build time by this script (and the list variant could be then modified at runtime by e.g. irq_connect_dynamic()).

sambhurst replied:
I've removed the max number of shared ISRs. Any number of ISRs can been shared.

microbuilder reviewed (commented):
Minor nitpick-ish comments are based on a quick browse through the PR .. I'll do a more thorough review and test when I'm back in the office next week.

microbuilder commented:
This feature will also need some sort of test. For example: [arm_irq_advanced_features](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/arch/arm/arm_irq_advanced_features)

galak commented:
would be nice to detect shared IRQs from devicetree :).

carlocaione reviewed (changes requested):
My biggest concern here is that this shared IRQ support is done leveraging the direct IRQs instead of regular IRQs (and the IRQ vector table instead of the ISR table). Now, direct IRQs are tricky, serviced in a special way and definitely harder to get them right.

So, why exactly not using the regular ISR table instead of the vector table?

andyross commented on scripts/build/gen_isr_tables.py:
```diff
@@ -14,6 +14,9 @@
 from elftools.elf.sections import SymbolTableSection
 
 ISR_FLAG_DIRECT = (1 << 0)
+ISR_FLAG_SHARE  = (1 << 1)
```
Unless there's a significant runtime savings for "unshared" IRQs that demands this flag, wouldn't it be nicer for the API to just work?  Register more than one ISR on the same interrupt and they all get called, no muss, no fuss.

Also avoids the significant[1] danger involved in duplicating a C API between headers and python code.

[1] Really.  I've absolutely tripped over issues like this.  It's a pretty bad footgun, because you really don't expect to need to look in python when you change a preprocessor symbol!

sambhurst replied:
I agree. I've removed the flags and any number of shared interrupts can be generated with IRQ_DIRECT_CONNECT

andyross reviewed (commented):
Some notes.  I like the feature, no so much the API.  What users want is almost certainly "shared interrupts just work" and not "use these special rules to register shared interrupts". 
 That's the use case, because if the drivers for those shared ISRs are tightly coupled and willing to write special code, they'd be able to just share an ISR between them anyway.

When app authors want "shared IRQs" it's because hardware designers messed up and the software is having to make things work by sharing an interrupt across multiple drivers/modules that **are not** related and **don't** want to have to do anything special.

How about CONFIG_KERNEL_SHARED_IRQ instead (and maybe an ARCH_SUPPORT_SHARED_IRQ flag), and when you have that set the kernel automatically knows that sw_isr_table[] entries are linked list records and that it needs to traverse them instead of just running the one it finds?

sambhurst commented:
> My biggest concern here is that this shared IRQ support is done leveraging the direct IRQs instead of regular IRQs (and the IRQ vector table instead of the ISR table). Now, direct IRQs are tricky, serviced in a special way and definitely harder to get them right.
> 
> So, why exactly not using the regular ISR table instead of the vector table?

The regular ISR table pass an argument to the ISR, which can't be used with the shared ISRs because they would need different arguments passed to them.

sambhurst commented:
> This feature will also need some sort of test. For example: [arm_irq_advanced_features](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/arch/arm/arm_irq_advanced_features)

@microbuilder I'm getting error while running the arm_irq_advanced_features tests. I'll upload my test for this PR when I can get the other tests to pass.

carlocaione commented:
> The regular ISR table pass an argument to the ISR, which can't be used with the shared ISRs because they would need different arguments passed to them.

Right, but having to force shared ISRs to have a NULL argument is a lesser evil here than force people to use the direct IRQs (that IMO are the real evil here).

Also I do not agree here:

> The shared_irq implementation doesn't seem very compatible with the current drivers and would require some code changes.

If the current implementation requires some changes I would prefer to have that fixed instead of introducing another incompatible way to do the same thing.

sambhurst commented:
> > The regular ISR table pass an argument to the ISR, which can't be used with the shared ISRs because they would need different arguments passed to them.
> 
> Right, but having to force shared ISRs to have a NULL argument is a lesser evil here than force people to use the direct IRQs (that IMO are the real evil here).
> 
> Also I do not agree here:
> 
> > The shared_irq implementation doesn't seem very compatible with the current drivers and would require some code changes.
> 
> If the current implementation requires some changes I would prefer to have that fixed instead of introducing another incompatible way to do the same thing.

@carlocaione I thought it would be confusing if a single ISR could be passed an argument but a shared ISR couldn't. Plus this could lead to bugs if an ISR is using an argument and it becomes NULL or changed when another ISR shares it's interrupt.

carlocaione commented:
> @carlocaione I thought it would be confusing if a single ISR could be passed an argument but a shared ISR couldn't.

Agree, but I don't see that a justification strong enough to make all the shared ISRs implemented using direct IRQs. I agree with @andyross here. Good the idea but the API and implementation needs a bit of rework.

sambhurst commented:
@carlocaione I've made the changes you requested. Please take a look when you have time. Thanks

carlocaione commented:
thanks but we are not still there. How am I supposed to use this? What's the use case here?

As @andyross wrote here https://github.com/zephyrproject-rtos/zephyr/pull/49427#pullrequestreview-1082636671 this should be transparent for the user.

Please write a test case for this where you have:

#### Test 1
you have two ISRs on two different IRQ lines. Each ISR is taking some kind of argument and doing something with that. Trigger the two IRQs and check that the two ISRs are called.

#### Test 2
Enable now some Kconfig option (`CONFIG_KERNEL_SHARED_IRQ` was suggested) and using the same two ISRs you now want to use the same IRQ line. Trigger both in some random order and check that everything works as expected. This should work without touching anything about the two ISRs.

gmarull commented on include/zephyr/irq.h:
```diff
@@ -35,14 +35,20 @@ extern "C" {
  * @param priority_p Interrupt priority.
  * @param isr_p Address of interrupt service routine.
- * @param isr_param_p Parameter passed to interrupt service routine.
+ * @param isr_param_p Parameter passed to interrupt service routine, unless the
+ *                    the IRQ is shared.
```
What is the reason to impose such limitation? What can I do in a multi-instance driver, then? See e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/adc/adc_gd32.c

carlocaione replied:
Probably inspired by my comment here https://github.com/zephyrproject-rtos/zephyr/pull/49427#issuecomment-1225303494 (but that came out wrongly as a suggestion to do that. My bad).

sambhurst commented:
> thanks but we are not still there. How am I supposed to use this? What's the use case here?
> 
> As @andyross wrote here [#49427 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/49427#pullrequestreview-1082636671) this should be transparent for the user.
> 
> Please write a test case for this where you have:
> 
> #### Test 1
> you have two ISRs on two different IRQ lines. Each ISR is taking some kind of argument and doing something with that. Trigger the two IRQs and check that the two ISRs are called.
> 
> #### Test 2
> Enable now some Kconfig option (`CONFIG_KERNEL_SHARED_IRQ` was suggested) and using the same two ISRs you now want to use the same IRQ line. Trigger both in some random order and check that everything works as expected. This should work without touching anything about the two ISRs.

@carlocaione I don't follow the second test case. If the 2 ISRs are using the arguments passed to them, those arguments will be set to NULL when the IRSs share the same IRQ, so those ISRs will not function properly when shared. Also, the CONFIG_KERNEL_SHARED_IRQ Kconfig has been removed. To share an IRQ, just invoke IRQ_CONNECT more than once.

carlocaione commented:
> @carlocaione I don't follow the second test case. If the 2 ISRs are using the arguments passed to them, those arguments will be set to NULL when the IRSs share the same IRQ, so those ISRs will not function properly when shared.

Right, but as suggested by @gmarull we need to be able to use the argument passed to the ISR (even when shared) to differentiate between different instances of multi-instance drivers.

sambhurst commented:
@carlocaione I've made the changes you requested but I still need to add tests. Please take another look when you have time.

aurel32 commented:
@sambhurst Thanks a lot for your work. I have just tested your patches (I haven't looked at their content in details yet) on an STM32G0B1, and they work great. I have tested that it's possible to share i2c2 and i2c3, and also usb and ucpd2. So I confirm they work without any code change beside enabling an additional device in the device tree. OTOH, that will allow to remove many existing workarounds in the drivers.

carlocaione commented on arch/arm/core/aarch32/Kconfig:
```diff
@@ -126,6 +126,13 @@ config COMPILER_ISA_THUMB2
 config NUM_IRQS
 	int
 
+config SHARED_ISR_TABLE
+	bool "Allow IRQS to share multiple ISRS"
+	default y
+	help
+	  Some ARM peripherals share an IRQ and this option enables the system
+	  to call all ISRs attached to a pending IRQ.
```
1. Why is this AArch32 specific?
2. This is confusing for someone that is used to the Zephyr nomenclature. The ISR table is something different than the vector table, yet this symbol is used to modify both.

gmarull replied:
> 1. Why is this AArch32 specific?

Agreed. In case of GD32VF103 (RISC-V) ADC0/1 also share the same IRQ line. I'd expect this option to be cross-arch (not necessary to provide all implementations, if arch specific quirks are needed)

carlocaione replied:
BTW, this is usually why (as it was suggested before) you want to have an `ARCH_HAS_*` symbol together with the actual symbol enabling the feature.

sambhurst replied:
> 1. Why is this AArch32 specific?

I only tested this PR on AArch32 boards and I thought the feature would be made available to other architectures by someone that could make sure it worked on said architecture.

> 2. This is confusing for someone that is used to the Zephyr nomenclature. The ISR table is something different than the vector table, yet this symbol is used to modify both.

I'll change the option name to ARCH_SUPPORT_SHARED_IRQ

carlocaione replied:
> I only tested this PR on AArch32 boards and I thought the feature would be made available to other architectures by someone that could make sure it worked on said architecture.

We have CI for that. And a bunch of QEMU boards as well.

sambhurst replied:
If all looks well with the AArch32 changes, I'll add the other architectures.

carlocaione commented on include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:
```diff
@@ -192,6 +192,9 @@ SECTIONS
 
 #include <zephyr/linker/kobject-rom.ld>
 
+	*(.shared_isr_args)
+	*(.shared_isr)
```
Is this going to work with `CONFIG_CMAKE_LINKER_GENERATOR=y` ?

sambhurst replied:
I've been testing with the cdc_acm_composite sample and enabling CONFIG_CMAKE_LINKER_GENERATOR, without any of my changes, generates build errors.

But setting CONFIG_CMAKE_LINKER_GENERATOR=y would likely fail with my changes too, so the generated linker file will need to include the .shared_isr and .shared_isr_args sections.

carlocaione commented on scripts/build/gen_isr_tables.py:
```diff
@@ -131,12 +131,65 @@ def parse_args():
 #endif
 """
 
+sh_isr_function = "void __attribute__((section(\".shared_isr\")))shared_isr_"
+sh_isr_args = "const size_t shared_isr_"
+sh_isr_args_section = " __attribute__((section(\".shared_isr_args\")))"
```
So, IIUC the sections are going to be populated only by this script. So before this script is run I would expect the sections to be empty. So, are you sure this is not going to modify (shifting) the addresses between the 1st and the 2nd linking stage?

Given that you provided no test I cannot verify it. Please, verify it by comparing the `linker_zephyr_pre{0,1}` files.

sambhurst replied:
No shifting occurs because I've placed the .shared_isr and .shared_isr_args sections at the end of all other sections.

I have some tests but I can't get the current tests,  [arm_irq_advanced_features](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/arch/arm/arm_irq_advanced_features), to pass. Is this the proper way to run them: twister -T tests/arch/arm/arm_irq_advanced_features/

ERROR   - mps2_an521 tests/arch/arm/arm_irq_advanced_features/arch.arm.irq_advanced_features.secure_fw FAILED: Unknown
ERROR   - mps2_an521 tests/arch/arm/arm_irq_advanced_features/arch.arm.irq_advanced_features FAILED: Unknown
ERROR   - mps2_an385 tests/arch/arm/arm_irq_advanced_features/arch.arm.irq_advanced_features FAILED: Unknown
ERROR   - qemu_cortex_m0  tests/arch/arm/arm_irq_advanced_features/arch.arm.irq_advanced_features FAILED: Unknown
ERROR   - mps2_an521_remote  tests/arch/arm/arm_irq_advanced_features/arch.arm.irq_advanced_features FAILED: Unknown
INFO    - Total complete:   86/  86  100%  skipped:   80, failed:    5

carlocaione commented on scripts/build/gen_isr_tables.py:
```diff
@@ -131,12 +131,65 @@ def parse_args():
             return key
     return None
 
+def write_shared_direct_isr_prototypes(fp, nv, shi):
+    for i in range(nv):
+        if len(shi[i]) > 0:
+            fp.write(sh_isr_function + "direct_" + str(i) + "(void);\n")
```
Are you sure this doesn't need a `__attribute__ ((interrupt))`. Did you verify in assembly what's generated by the compiler?

sambhurst replied:
I tested the direct ISR and it worked but I agree with you about adding __attribute__ ((interrupt))

carlocaione commented on scripts/build/gen_isr_tables.py:
```diff
@@ -131,12 +131,65 @@ def parse_args():
+def write_shared_isr_args(fp, nv, shi):
+    for i in range(nv):
+        if len(shi[i]) > 0:
+            fp.write("const size_t shared_isr_" + str(i) + "_args[" +
```
Why args are now `size_t`?

sambhurst replied:
I thought size_t would be more portable when this feature is enabled on other architectures.

carlocaione replied:
Please stick to `(const void *)`

carlocaione commented on include/zephyr/arch/arm/aarch32/cortex_a_r/scripts/linker.ld:
```diff
@@ -202,6 +202,8 @@ SECTIONS
 
 #include <zephyr/linker/kobject-rom.ld>
 
+        *(.shared_isr_args)
+        *(.shared_isr)
```
How did you decide where to put those symbols? Some platforms do not have much freedom in that regard. See for example 8c4f98d263865d00199fc9b5f7870a0f134c634b.

sambhurst replied:
I placed them after all other sections to avoid shifting of the address.

carlocaione replied:
This is wrong for several reasons. A couple here.
1. These could be the latest sections when using this vanilla linker script. But some boards can add some custom sections after, so those are going to be shifted.
2. Some platforms require the ISR to be in some specific sections or delimited by special symbols (see 8c4f98d263865d00199fc9b5f7870a0f134c634b).

The whole point of having `arch/common/isr_tables.c` is indeed having to reserve the space in advance so that the symbols are not going to be shifted. Otherwise we could have moved the location of the tables in the first place.

sambhurst replied:
Thanks for your patience, I'm new to this part of Zephyr. I've moved the shared ISRs and arguments to the .last_section. If this is still incorrect, please advise on the proper way to implement this.

carlocaione reviewed (changes requested):
Besides some very fundamental concerns I still need to see this in action using some tests.

sambhurst committed:
tests: interrupt: Add tests for shared irqs

Add test case of shared interrupts. This tests
that IRQs are shared when IRQ_CONNECT is called
multiple times on the same IRQ with different
ISR handlers.

Signed-off-by: Sam Hurst <sbh1187@gmail.com>

carlocaione commented on arch/arm/core/aarch32/Kconfig:
```diff
@@ -126,6 +126,13 @@ config COMPILER_ISA_THUMB2
 config NUM_IRQS
 	int
 
+config ARCH_SUPPORT_SHARED_IRQ
+	bool "Allow IRQs to share multiple ISRs"
+	default n
+	help
+	  Some ARM peripherals share an IRQ and this option enables the system
+	  to call all ISRs attached to a pending IRQ.
+
```
This still needs to go in `arch/Kconfig` not here. Also, better using `ARCH_HAS`.

sambhurst replied:
I've moved option to arch/Kconfig and placed the option in the "Interrupt Configuration" section, named ARCH_SW_ISR_TABLE_SHARED_IRQ

carlocaione commented on scripts/build/gen_isr_tables.py:
```diff
@@ -131,12 +131,65 @@ def parse_args():
+def write_shared_direct_isr_prototypes(fp, nv, shi):
+    for i in range(nv):
+        if len(shi[i]) > 0:
+            fp.write("__attribute__((interrupt(\"IRQ\"))) " + sh_isr_function + "direct_" + str(i) + "(void);\n")
```
this is definitely not portable.

carlocaione commented on scripts/build/gen_isr_tables.py:
```diff
@@ -172,14 +225,20 @@ def write_address_irq_vector_table(fp, vt, nv):
 
     nv = intlist["num_vectors"]
 
+    if "CONFIG_ARCH_SUPPORT_SHARED_IRQ" in syms:
```
The `CONFIG_ARCH_HAS` symbols are not used like this. You have `CONFIG_ARCH_HAS_*` enabled by the architecture and then you have a different symbol `CONFIG_WHATEVER` depending on `CONFIG_ARCH_HAS*` enabled by the SoC / board. Your check must be on `CONFIG_WHATEVER`, not on `CONFIG_ARCH_HAS_*`

sambhurst replied:
I've made change

carlocaione commented on tests/arch/arm/arm_irq_advanced_features/prj.conf:
```diff
@@ -3,3 +3,4 @@ CONFIG_DYNAMIC_INTERRUPTS=y
 CONFIG_DYNAMIC_DIRECT_INTERRUPTS=y
 CONFIG_ZERO_LATENCY_IRQS=y
 CONFIG_ZTEST_NEW_API=y
+CONFIG_ARCH_SUPPORT_SHARED_IRQ=y
```
This must be set by the architecture.

sambhurst replied:
removed

carlocaione commented on tests/arch/arm/arm_irq_advanced_features/src/arm_shared_isrs.c:
```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2022 The Chromium OS Authors
```
are you one of them?

sambhurst replied:
Yes

carlocaione commented on tests/arch/arm/arm_irq_advanced_features/src/arm_shared_isrs.c:
```diff
@@ -0,0 +1,105 @@
+
+#include <zephyr/ztest.h>
+#include <zephyr/arch/cpu.h>
+#include <zephyr/arch/arm/aarch32/cortex_m/cmsis.h>
```
again. This must not be aarch32 specific.

sambhurst replied:
The other tests in this directory are also aarch32 specific. Should this test be implemented in another place?

carlocaione replied:
yes, this should be moved to `tests/kernel/interrupt/`

andyross reviewed (commented):
Still not understanding why this is being implemented in terms of ARM-only "direct" interrupts and not the portable layer.  Direct interrupts exist only as a performance optimization, which you're giving up here anyway by handling multiple function pointers in the same ISR.

Why not do this in a portable way on top of the existing _sw_isr_tables mechanism, etc...?

sambhurst commented:
> Still not understanding why this is being implemented in terms of ARM-only "direct" interrupts and not the portable layer. Direct interrupts exist only as a performance optimization, which you're giving up here anyway by handling multiple function pointers in the same ISR.
> 
> Why not do this in a portable way on top of the existing _sw_isr_tables mechanism, etc...?

I've removed the shared direct ISRs.

andyross reviewed (dismissed):
I'm out of complaints, thanks for your patience.  As a parting nitpick: my gut says this is too much python and not enough C.  Generating the indirected functions at build time seems more complicated (and probably more bytes) than just putting a "next" pointer into the ISR table entry.  But that's mostly a style thing.

carlocaione commented on arch/Kconfig:
```diff
@@ -454,6 +454,13 @@ config ARCH_SW_ISR_TABLE_ALIGN
 	  to be aligned to architecture specific size.  The default
 	  size is 0 for no alignment.
 
+config ARCH_SW_ISR_TABLE_SHARED_IRQ
```
why this is not `ARCH_HAS_*`?

sambhurst replied:
Done

carlocaione commented on scripts/build/gen_isr_tables.py:
```diff
@@ -131,12 +131,51 @@ def parse_args():
 """
 
+sh_isr_function = "__attribute__((section(\".shared_isr\"))) void shared_isr_"
+sh_isr_args = "const uint32_t shared_isr_"
```
what about 64bit platforms?

sambhurst replied:
I've changed the uint32_t to size_t and tested it on arm32 and arm64. On arm64, I passed in a 64bit value as the argument to the ISR and verified there wasn't a build error and that the generated isr_tables.c had the correct value

carlocaione commented on scripts/build/gen_isr_tables.py:
```diff
@@ -131,12 +131,51 @@ def parse_args():
+            idx = 0
+            for isr in shi[i]:
+                func = str(hex(isr[1]))
+                fp.write("\t((ISR)" + func + ")((void *)((uint32_t *)args)[" +
```
why `(void *)(uint32_t *)`?

sambhurst replied:
This is been changed to (void *)(size_t *). The initial (size_t *) cast is used so the compiler can index the args array and the (void *) cast is to match the type expected by the ISR.

carlocaione commented on arch/arm/core/aarch32/Kconfig:
```diff
@@ -126,6 +129,14 @@ config COMPILER_ISA_THUMB2
 config NUM_IRQS
 	int
 
+config SUPPORT_SHARED_IRQ
```
this is (again) not something AArch32 specific. This should be moved to somewhere accessible by all the architectures.

sambhurst replied:
Removed

carlocaione commented on include/zephyr/arch/arm/aarch32/cortex_a_r/scripts/linker.ld:
```diff
@@ -386,8 +386,10 @@ GROUP_END(OCM)
 
     /* Must be last in romable region */
-    SECTION_PROLOGUE(.last_section,(NOLOAD),)
+    SECTION_PROLOGUE(.last_section,,)
     {
+	*(.shared_isr_args)
+	*(.shared_isr)
```
I'm very very skeptical about this, so you must really convince me that this is the correct way to do this.

1. This is working with `CONFIG_CMAKE_LINKER_GENERATOR=y ` ?
2. What happens to the platforms that do not have a `.last_section`?
3. Why this is (again) AArch32 specific?

sambhurst replied:
@carlocaione I'm working on an STM32 project that requires both USB and UCPD but they share IRQ 8. So as it stands, I can't use those peripheral together and my project is on hold. I was mainly focusing on AArch32 because that's the architecture of my microcontroller but I planned to add other architectures to this PR once I got approved changes for the AArch32.

I'm relatively new to Zephyr and have mostly contributed drivers and subsystems but not fundamental changes such as those in this PR. Could you advise on the correct way to implement shared ISRs in Zephyr.

andyross replied:
FWIW: I'm fine if this is supported by only one arch to start with as long as the API is clean.  The code to manage and interpret the ISR table should be generic and documented, but the instructions that actually invoke the pointers are all in per-architecture interrupt entry layers.

carlocaione replied:
I'm fine with having this supported by only one arch to start with, but some trivial (possibly) changes can make this PR easily adoptable by other architectures as well.

Instead right now we have:
- AArch32 specific test while this should really be a general test for all the architectures supporting this because it's not relying on anything special to AArch32. We can of course enable this feature and test only for AArch32 at the beginning.
- The whole mechanism relies on a change to AArch32 linker scripts while we should really make this generic (see 3a48365babbdacda4547a2076061592912b6e474 for inspiration and in general the whole #46769 where we tried to decouple the arch linker scripts from whatever it is generated by the `gen_isr_tables.py` script)
- `CONFIG_CMAKE_LINKER_GENERATOR` is not correctly supported yet. When compiling the test we get:
```
warning: orphan section `.shared_isr' from `zephyr/CMakeFiles/zephyr_final.dir/isr_tables.c.obj' being placed in section `.shared_isr'
warning: orphan section `.shared_isr_args' from `zephyr/CMakeFiles/zephyr_final.dir/isr_tables.c.obj' being placed in section `.shared_isr_args'
```
In general the changes to the PR are slowly converging but IMO this is still a bit of work to do but it's really matter of moving to a standalone test and adding a bunch of `zephyr_linker_sources_ifdef` instead of touching the linker scripts.

sambhurst replied:
Thanks for the detailed explanation. I've uploaded a new PR that hopefully moves things in the right direction. General interrupt tests still need to be written. I have not found a way to set a pending interrupt in the tests that works like NVIC_SetPendingIRQ.

sambhurst replied:
Also, setting CONFIG_CMAKE_LINKER_GENERATOR=y does not generate build errors.

carlocaione reviewed (changes requested):
Sorry, I'm going to -1 this until this support is made general and not only AArch32 specific.

carlocaione commented on tests/arch/arm/arm_irq_advanced_features/src/arm_shared_irq.c:
```diff
@@ -0,0 +1,112 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/ztest.h>
+#include <zephyr/arch/cpu.h>
+#include <zephyr/arch/arm/aarch32/cortex_m/cmsis.h>
```
This all test should be moved to `tests/kernel/interrupt` and properly `ifdef`-fed to work only for AArch32 for now.

sambhurst replied:
I moved my test to tests/kernel/interrupt and the AArch32 tests can be moved in a separate PR.

carlocaione commented on arch/common/shared_irq.ld:
```diff
@@ -0,0 +1,15 @@
+
+_shared_irq_start = .;
+
+KEEP(*(.shared_irq))
+KEEP(*(.shared_irq_arg))
```
Still something wrong for me with `CONFIG_CMAKE_LINKER_GENERATOR`.
```
$ west build -b qemu_cortex_m3 tests/arch/arm/arm_irq_advanced_features --pristine -- -DCONFIG_CMAKE_LINKER_GENERATOR=y

...
warning: orphan section `.shared_irq' from `zephyr/CMakeFiles/zephyr_final.dir/isr_tables.c.obj' being placed in section `.shared_irq'
warning: orphan section `.shared_irq_arg' from `zephyr/CMakeFiles/zephyr_final.dir/isr_tables.c.obj' being placed in section `.shared_irq_arg'

```

sambhurst replied:
I've made an addition to the PR that works with CONFIG_CMAKE_LINKER_GENERATOR

```
$ west build -b qemu_cortex_m3 tests/arch/arm/arm_irq_advanced_features --pristine -- -DCONFIG_CMAKE_LINKER_GENERATOR=y

...
[150/150] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       20212 B       256 KB      7.71%
             RAM:        5719 B        64 KB      8.73%
        IDT_LIST:          0 GB         2 KB      0.00%
```

sambhurst commented on arch/common/shared_irq.ld:
```diff
@@ -0,0 +1,15 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* #if defined(CONFIG_CPU_CORTEX_M_HAS_VTOR) */
```
Done

sambhurst commented on scripts/build/gen_isr_tables.py:
```diff
@@ -131,12 +131,51 @@ def parse_args():
 """
 
+sh_isr_function = "__attribute__((section(\".shared_irq\"))) void shared_irq_"
+sh_isr_args = "const size_t shared_irq_"
```
Done

sambhurst commented:
@carlocaione could you take another look? Thanks

carlocaione commented on tests/kernel/interrupt/testcase.yaml:
```diff
@@ -11,3 +11,4 @@ tests:
     filter: not CONFIG_TRUSTED_EXECUTION_NONSECURE
     extra_configs:
       - CONFIG_CMAKE_LINKER_GENERATOR=y
+      - CONFIG_ISR_SHARE_IRQ=y
```
Why are you testing this only for the `linker_generator` subtest? This should probably go in the `prj.conf` file instead if you want to test both cases (`CONFIG_CMAKE_LINKER_GENERATOR` and `!CONFIG_CMAKE_LINKER_GENERATOR`).

sambhurst replied:
Moved to prj.conf

carlocaione commented on tests/kernel/interrupt/CMakeLists.txt:
```diff
@@ -15,5 +15,8 @@ target_sources(app PRIVATE
     src/nested_irq.c
     )
 
+if(CONFIG_CPU_CORTEX_M)
+  target_sources_ifdef(CONFIG_ISR_SHARE_IRQ app PRIVATE src/shared_irq.c)
+endif()
```
This is not the correct way to implement this. See my later comment.

sambhurst replied:
The guard has been removed.

carlocaione commented on tests/kernel/interrupt/src/shared_irq.c:
```diff
@@ -0,0 +1,119 @@
+	irq_enable(ISR3_4_5_OFFSET);
+	irq_enable(ISR6_OFFSET);
+
+	NVIC_SetPendingIRQ(ISR0_OFFSET);
+	NVIC_SetPendingIRQ(ISR1_2_OFFSET);
+	NVIC_SetPendingIRQ(ISR3_4_5_OFFSET);
+	NVIC_SetPendingIRQ(ISR6_OFFSET);
```
What happens if this test must be reused for a platform that does not support NVIC? This test should be generic (again), not AArch32 specific. Please, see for example `nested_irq.c` to see the correct way to implement this (this should be guarded by a `#if defined(CONFIG_CPU_CORTEX_M)` basically).

sambhurst replied:
This has been changed.

carlocaione commented on arch/Kconfig:
```diff
@@ -34,6 +34,7 @@ config ARM
 	# is really only necessary for Cortex-M with ARM MPU!
 	select GEN_PRIV_STACKS
 	select ARCH_HAS_THREAD_LOCAL_STORAGE if CPU_AARCH32_CORTEX_R || CPU_CORTEX_M || CPU_AARCH32_CORTEX_A
+	select ARCH_HAS_SW_ISR_TABLE_SHARED_IRQ
```
besides the test, there is anything else preventing this to be enabled for other architectures as well?

sambhurst replied:
I enabled this on the other architectures and some of the devices fails the shared_isr test.

sambhurst replied:
Could we release this feature on a per architecture basis? I'm not sure why this is working for the cortex_m and failing on the other architectures.

carlocaione replied:
I mean that's concerning right? We are introducing a new wide features that for some reason is only working one one architecture and failing on all the other ones. At the very minimum we should have an idea of why this is the case.

Anyway, not giving my -1 because I have already blocked this for too long. Not giving +1 either.

sambhurst replied:
I agree with you but I don't have the expertise to debug the other architectures. How would you recommend I proceed?

andyross reviewed (dismissed):
refresh +1, looks like this is getting hot again

carlocaione reviewed (changes requested):
While replying to https://github.com/zephyrproject-rtos/zephyr/pull/49427#discussion_r998423786 and investigating why this was failing on ARM64, I realized that this PR has some serious issues that need to be addressed.

In particular on ARM64 this is failing because the generated code for the shared ISRs is not even placed in the text section.

That is important on ARM64 because it is the only region that is MMU-mapped to be EXecutable, so this is faulting.

We need to at least have the ISR code in the text section.

Cherish-forever commented:
@carlocaione  Please review.

tejlmand commented on include/zephyr/arch/arm/aarch32/cortex_m/scripts/linker.ld:
```diff
@@ -135,6 +135,8 @@ SECTIONS
 	*(.text)
 	*(".text.*")
 	*(".TEXT.*")
+        *(.shared_irq)
+        *(.shared_irq_arg)
```
this code together with:
https://github.com/zephyrproject-rtos/zephyr/blob/ffd7107810322e5de7f86b37ed88b2533cc0acdf/arch/common/shared_irq.ld#L11-L12

means you now get this twice in the linker script, in two output sections, like this:
```
 60     text :                                                                                                                                                                      
 61  {                                                                                                                                                                              
 62  __text_region_start = .;                                                                                                                                                       
 63  *(.text)                                                                                                                                                                       
 64  *(".text.*")                                                                                                                                                                   
 65  *(".TEXT.*")                                                                                                                                                                   
 66         *(.shared_irq)                                                                                                                                                          
 67         *(.shared_irq_arg)                                                                                                                                                      
 68  *(.gnu.linkonce.t.*)                 
....
180     rodata :                                                                                        
181  {                                                                                                  
182  *(.rodata)                                                                                         
183  *(".rodata.*")                                                                                     
184  *(.gnu.linkonce.r.*)                                                                               
185 . = ALIGN(4);                                                                                       
186 _shared_irq_start = .;                                                                              
187 KEEP(*(.shared_irq))                                                                                
188 KEEP(*(.shared_irq_arg))                                                                            
189 _shared_irq_end = .;         
...
```

that doesn't seem right.

tejlmand commented on include/zephyr/arch/arm64/scripts/linker.ld:
```diff
@@ -111,6 +111,8 @@ SECTIONS
 
         *(.text)
         *(".text.*")
+        *(.shared_irq)
+        *(.shared_irq_arg)
```
similar to other comment.

tejlmand commented on kernel/Kconfig:
```diff
@@ -180,6 +180,13 @@ config IDLE_STACK_SIZE
 	  logging (e.g. logging when the system goes to sleep), the idle thread
 	  may need more stack space than the default value.
 
+config ISR_SHARE_IRQ
+	bool "Allow IRQs to share multiple ISRs"
+	depends on ARCH_HAS_SW_ISR_TABLE_SHARED_IRQ
+	help
+	  Some peripherals share an IRQ and this option enables the system
+	  to call all ISRs attached to a pending IRQ.
```
Why is this not placed in the `Interrupt Configuration` menu here:
https://github.com/zephyrproject-rtos/zephyr/blob/460ac477c1e9fcf36cee82c4d3c0571476a4bca2/arch/Kconfig#L380-L389

Secondly, should this setting be dependent on `GEN_ISR_TABLES` ?

tejlmand commented on arch/common/CMakeLists.txt:
```diff
@@ -75,3 +75,5 @@ if(CONFIG_COVERAGE)
 
 zephyr_sources_ifdef(CONFIG_SEMIHOST semihost.c)
+
+zephyr_linker_sources(RODATA SORT_KEY default shared_irq.ld)
```
between stage 0 and stage 1 linking, we can resize, but between stage `final-1` (stage 1 in this case) and final linking stage we should avoid resizing due to the invocation of `gen_handles` and `gperf` between those two stages.

tejlmand commented on tests/kernel/interrupt/prj.conf:
```diff
@@ -4,3 +4,4 @@ CONFIG_DYNAMIC_INTERRUPTS=y
 CONFIG_MP_MAX_NUM_CPUS=1
 CONFIG_THREAD_STACK_INFO=y
 CONFIG_IRQ_OFFLOAD=y
+CONFIG_ISR_SHARE_IRQ=y
```
please keep the existing test cases to test what they already do today, that is without introducing `CONFIG_ISR_SHARE_IRQ=y`.

Instead add a new test case here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/kernel/interrupt/testcase.yaml
For example: `arch.interrupt.shared:`

where you can set `CONFIG_ISR_SHARE_IRQ=y`.

That will also allow you to limit testing to archs you know are supported.

sambhurst commented:
@tejlmand could you take another look? Thanks

Cherish-Gww commented:
@sambhurst 
I found that in zephyr_pre1.elf stage struct device address is different with zephyr.elf
here is zephyr_pre1.elf:
  1685: 0800be44    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_43
  1723: 0800bea4    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_23
  1919: 0800beec    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_53
  1974: 0800be8c    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_73
  2064: 0800be5c    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_75
  2154: 0800be2c    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_8
  2228: 0800bf64    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_103
  2256: 0800bf34    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_99
  2321: 0800bebc    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_27
  2417: 0800be74    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_74
  2463: 0800bed4    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_25
  2575: 0800bf4c    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_34
  2592: 0800bf1c    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_102
  2614: 0800bf04    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_13
  
  and this is zephyr.elf
    1673: 0800be7c    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_43
  1711: 0800bedc    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_23
  1907: 0800bf24    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_53
  1962: 0800bec4    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_73
  2052: 0800be94    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_75
  2142: 0800be64    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_8
  2216: 0800bf9c    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_103
  2244: 0800bf6c    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_99
  2309: 0800bef4    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_27
  2406: 0800beac    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_74
  2453: 0800bf0c    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_25
  2565: 0800bf84    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_34
  2582: 0800bf54    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_102
  2605: 0800bf3c    24 OBJECT  GLOBAL DEFAULT    5 __device_dts_ord_13
  
  struct device address has been changed, shared interrupt point to zephyr.pre1.elf address.

Cherish-Gww commented on cmake/linker_script/arm/linker.cmake:
```diff
@@ -83,6 +83,8 @@ zephyr_linker_section_configure(SECTION .text INPUT ".glue_7t")
 zephyr_linker_section_configure(SECTION .text INPUT ".glue_7")
 zephyr_linker_section_configure(SECTION .text INPUT ".vfp11_veneer")
 zephyr_linker_section_configure(SECTION .text INPUT ".v4_bx")
+zephyr_linker_section_configure(SECTION .text INPUT ".shared_irq")
```
.text section is before .device, in final link, it will change .device section address.

Cherish-Gww commented on include/zephyr/arch/arm/aarch32/cortex_a_r/scripts/linker.ld:
```diff
@@ -141,6 +141,8 @@ SECTIONS
 
         *(.text)
         *(".text.*")
+        *(.shared_irq)
```
Move .shared_irq and shared_irq_arg to .sw_isr_table will not influence the .device address

carlocaione replied:
this is not possible. The code dealing with the shared interrupt must be somewhere in the `.text` section, see https://github.com/zephyrproject-rtos/zephyr/pull/49427#pullrequestreview-1146218087. If not, must be in a section with the correct settings for eXecution for MMU / MPU.

Cherish-Gww replied:
> this is not possible. The code dealing with the shared interrupt must be somewhere in the `.text` section, see [#49427 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/49427#pullrequestreview-1146218087). If not, must be in a section with the correct settings for eXecution for MMU / MPU.

Put text section blew device is it possible?

carlocaione replied:
Not quite sure I understood the question. If you are asking whether the current placement of the `shared_irq` code in the `text` section is correct then it is not. This was a problem already seen in the past, see https://github.com/zephyrproject-rtos/zephyr/pull/49427#pullrequestreview-1095642300

carlocaione replied:
We need to have:
- The `shared_irq` code in a section that is mapped as eXecutable (`.text` is a natural choice, but I'm open to anything really) because the script is generating code at build time that (of course) needs to be eXecuted.
- As pointed out by @tejlmand the shifting addresses are allowed only between stage 0 and 1. Having shifting addresses between other linking stages could be problematic.

Cherish-Gww replied:
> Not quite sure I understood the question. If you are asking whether the current placement of the `shared_irq` code in the `text` section is correct then it is not. The shifting of the addresses was a problem already seen in the past, see [#49427 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/49427#pullrequestreview-1095642300)

maybe we can try link device section first, then link shared irq in text section.

Cherish-Gww replied:
> We need to have:
> 
> * The `shared_irq` code in a section that is mapped as eXecutable (`.text` is a natural choice, but I'm open to anything really) because the script is generating code at build time that (of course) needs to be eXecuted.
> * As pointed out by @tejlmand the shifting addresses are allowed only between stage 0 and 1. Having shifting addresses between other linking stages could be problematic.

Here's my opinion:
>  The code does not have to be in the text section.
> sw_isr_table link stage is blew text and device, so that function in text section address and device in device section address will not be change.

so put shared_irq and shared_irq_arg to sw_isr_table will not effect.

erwango commented:
@sambhurst For my information, do you plan to complete this PR ?

sambhurst commented:
@erwango I'd like to finish this PR but there doesn't seem to be a general agreement on how this feature should be implemented. To simplify the effort, I proposed to just implement and test it for Arm Cortex devices and later roll it out to other devices, but this proposal wasn't well received.  So, unless we can arrive at a consensus, we can close this PR.

Cherish-Gww commented:
> @erwango I'd like to finish this PR but there doesn't seem to be a general agreement on how this feature should be implemented. To simplify the effort, I proposed to just implement and test it for Arm Cortex devices and later roll it out to other devices, but this proposal wasn't well received. So, unless we can arrive at a consensus, we can close this PR.

Did this issue come up in your test? and my solution is right?
https://github.com/zephyrproject-rtos/zephyr/pull/49427#issuecomment-1321722430

aurel32 commented:
I would like to get shared IRQ working, it's important for some MCU. For what I understood the remaining issues are:
- Device addresses change between linking stages due to shared IRQ handlers being added for the text section
- The shared IRQ handlers must be in the text section as they need to be executable. This is the root of the issues on some other architectures (e.g. arm64)

Both issues seems to contradict themselves, therefore I have just tried an alternative way with a common shared irq handler getting passed the data, i.e. both argument and function. For that I reused the existing `_isr_table_entry` struct, but that can be changed. The common shared irq handler is mark as `KEEP` in the linker file to avoid moving symbols between linking stages, as it is not referenced in the first linking stage. It seems to work fine on a Cortex-M CPU. My WIP code is there:  https://github.com/aurel32/zephyr/commit/fedbe2b4d571700d0990dc24b0989f8bbc584ff7

Before continuing in that direction, and then getting things working on at least a second architecture, I would like to know if that strategy is acceptable? Things like naming, comments and so on, can probably be improved, but before investing time on that I would like to get some feedback about the big picture.

Cherish-Gww commented:
> * The shared IRQ handlers must be in the text section as they need to be executable. This is the root of the issues on some other architectures (e.g. arm64)

I agree with your patch,  
an independent section needs to declare its executable permissions in the MPU, 
This may be the reason for solving the above problem on arm64.

carlocaione commented:
> I would like to know if that strategy is acceptable?

Please, open a (different) PR with your code so people can have a chance to take a look and we can check if CI is happy with this solution as well.

Cherish-Gww commented:
> > I would like to know if that strategy is acceptable?
> 
> Please, open a (different) PR with your code so people can have a chance to take a look and we can check if CI is happy with this solution as well.
@carlocaione @aurel32 
I did not submit a PR, here are my thoughts, I'm just a little white, please don't mock me:

`   
    #include <zephyr/linker/common-rom.ld>        /* Move common-rom.ld before text section linking */
 
    SECTION_PROLOGUE(_TEXT_SECTION_NAME,,)
	{
	__text_region_start = .;
#include <zephyr/linker/kobject-text.ld>
	*(.text)
	*(".text.*")
	*(".TEXT.*")
	*(.gnu.linkonce.t.*)
        KEEP(*(.shared_irq))`
        
I have tried on qemu_cortex_m3 board, hello_world runs well.

the elf section header is:
`Section Headers:
  [Nr] Name              Type            Addr     Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            00000000 000000 000000 00      0   0  0
  [ 1] rom_start         PROGBITS        00000000 0000d4 0000ec 00 WAX  0   0  4
  [ 2] initlevel         PROGBITS        000000ec 0001c0 000038 00   A  0   0  4
  [ 3] devices           PROGBITS        00000124 0001f8 000048 00   A  0   0  4                        /* struct device section */
  [ 4] sw_isr_table      PROGBITS        0000016c 000240 000158 00  WA  0   0  4
  [ 5] device_handles    PROGBITS        000002c4 000398 000012 00   A  0   0  2
  [ 6] log_const_se[...] PROGBITS        000002d8 0003ac 000010 00   A  0   0  4
  [ 7] text              PROGBITS        000002e8 0003bc 001e20 00  AX  0   0  4
  [ 8] .ARM.exidx        ARM_EXIDX       00002108 0021dc 000008 00  AL  7   0  4                 /* the text section */
  [ 9] rodata            PROGBITS        00002110 0021e4 0000ec 00   A  0   0  4
  [10] .ramfunc          PROGBITS        20000000 0022fe 000000 00   W  0   0  1
  [11] datas             PROGBITS        20000000 0022d0 000024 00  WA  0   0  4
  [12] device_states     PROGBITS        20000024 0022f4 000006 00  WA  0   0  1
  [13] bss               NOBITS          20000030 002300 00023e 00  WA  0   0  8
  [14] noinit            NOBITS          20000270 002300 000d00 00  WA  0   0  8
  [15] .comment          PROGBITS        00000000 0022fe 000020 01  MS  0   0  1
  [16] .debug_aranges    PROGBITS        00000000 002320 000ad0 00      0   0  8
  [17] .debug_info       PROGBITS        00000000 002df0 026c4c 00      0   0  1
  [18] .debug_abbrev     PROGBITS        00000000 029a3c 005a98 00      0   0  1
  [19] .debug_line       PROGBITS        00000000 02f4d4 00c9e4 00      0   0  1
  [20] .debug_frame      PROGBITS        00000000 03beb8 001618 00      0   0  4
  [21] .debug_str        PROGBITS        00000000 03d4d0 00452c 01  MS  0   0  1
  [22] .debug_loc        PROGBITS        00000000 0419fc 00ba68 00      0   0  1
  [23] .debug_ranges     PROGBITS        00000000 04d468 001f90 00      0   0  8
  [24] .ARM.attributes   ARM_ATTRIBUTES  00000000 04f3f8 00002b 00      0   0  1
  [25] .last_section     PROGBITS        00002226 0022fa 000004 00   A  0   0  1
  [26] .symtab           SYMTAB          00000000 04f424 0033c0 10     27 364  4
  [27] .strtab           STRTAB          00000000 0527e4 00300f 00      0   0  1
  [28] .shstrtab         STRTAB          00000000 0557f3 000132 00      0   0  1
Key to Flags:
`

aurel32 commented:
> > > I would like to know if that strategy is acceptable?
> > 
> > 
> > Please, open a (different) PR with your code so people can have a chance to take a look and we can check if CI is happy with this solution as well.
> 
> @carlocaione @aurel32 I did not submit a PR, here are my thoughts, I'm just a little white, please don't mock me:
> 
> ```
>     #include <zephyr/linker/common-rom.ld>        /* Move common-rom.ld before text section linking */
>  
>     SECTION_PROLOGUE(_TEXT_SECTION_NAME,,)
> 	{
> 	__text_region_start = .;
> #include <zephyr/linker/kobject-text.ld>
> 	*(.text)
> 	*(".text.*")
> 	*(".TEXT.*")
> 	*(.gnu.linkonce.t.*)
>         KEEP(*(.shared_irq))
> ```
> 
> I have tried on qemu_cortex_m3 board, hello_world runs well.

Thanks for testing ! On my side I forgot to mention here that I opened a PR as requested, with the test passing on all tested platforms, that is: qemu_cortex_m0 qemu_cortex_m3 qemu_cortex_a9 qemu_cortex_a53 qemu_cortex_a53_smp

Cherish-Gww commented:
@aurel32 
Your solution is more reasonable

sambhurst commented:
@aurel32 Thanks for picking this up.
